Errors out if json machine output has not been implemented

cdrage opened the pull request:
This PR:    
    
 - Moves annotations to `&cobra.Command` struct creation rather than    
 using the command to add it    
 - Errors out if `-o json` has been passed, but json support has not    
 been added    
 - Hides the `-o json` command if it's not impemented    
 - Solves the issue of https://github.com/openshift/odo/issues/2031    
 - Changes "Global Flags" to "Additional Flags"
  
Closes https://github.com/openshift/odo/issues/2031

cdrage commented:
@amitkrout if you could please review that'd be great.

Keep in mind that this does not remove `-o` from the help menu, but rather, it will error out if the user passes in `-o json` to a command that does not have the functionality implemented.

mik-dass commented:
```
odo storage list --context /tmp/800490314 -o json]
[odo]  ✗  Machine readable output is not yet implemented for this command
```
@cdrage storage list is missing the required changes

mik-dass commented:
@cdrage seems like odo version also needs the changes too
```
odo sub component command tests Generic machine readable output tests 
  Command should fail if json is non-existent for a command<`2`>
  /home/travis/gopath/src/github.com/openshift/odo/tests/integration/component.go:44
Created dir: /tmp/602711673
Running odo with args [odo version -o json]
[odo]  ✗  Machine readable output is not yet implemented for this command
Deleting dir: /tmp/602711673
```

mik-dass commented on tests/integration/component.go:
```diff
@@ -39,6 +39,18 @@ func componentTests(args ...string) {
+			// `odo version` does not have json implemented, and thus we will use this
+			// as a test
+			output := helper.CmdShouldFail("odo", "version", "-o", "json")
+			Expect(output).To(ContainSubstring("Machine readable output is not implemented for this command"))
```
```
Expected
        <string>:  ✗  Machine readable output is not yet implemented for this command
        
    to contain substring
        <string>: Machine readable output is not implemented for this command
```
`is not yet implemented`

cdrage commented:
/test v4.1-integration

mik-dass commented:
```
 <*os.PathError | 0xc0003060c0>: {Op: "chdir", Path: "", Err: 0x2}
        chdir : no such file or directory
```
/retest

mohammedzee1000 commented on pkg/odo/genericclioptions/runnable.go:
```diff
@@ -30,15 +30,29 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) {
+	machineOutput := cmd.Annotations["machineoutput"]
+
+	if flagChanged && outputFlag.Value.String() != "json" {
 		log.Error("Please input a valid output format for -o, available format: json")
```
All these (log.Error + exit 1) could possibly be replaced by a [log.Fatal like the standard log package](https://golang.org/pkg/log/#Fatal) maybe. Anycase more of an implementation detail :)

cdrage replied:
Yeah, we use `log.Error` instead since we're able to modify the output to something more pretty :)

kadel commented:
I can still see that the help messages are showing the `-o` flag even for commands where it is not implemented.
@cdrage I taught that you mentioned that we can hide it from the output, so it is not shown for the commands where it is not implemented.

cdrage commented:
@kadel Yes, this is only one half of the issue. This errors out if it's been passed in, but I have not yet updated it to hide the actual flag.. I'll set this as WIP for now.

kadel commented:
> @kadel Yes, this is only one half of the issue. This errors out if it's been passed in, but I have not yet updated it to hide the actual flag.. I'll set this as WIP for now.

We can split it into multiple PRs if you want. I just wanted to make sure that we forgot this part, as it is also really important.

cdrage commented:
Updated the PR! Ready for review @kadel and @girishramnani 

I've updated the PR's description and implementing hiding the command.

Global Flags has also been renamed to Additional Flags.

See below for the changes:

```sh
github.com/openshift/odo  remove-json-from-some-commands ✔                                                                                                                                                                                                                 13m  ⍉
▶ ./odo version --help
Print the client version information

Usage:
  odo version [flags]

Examples:
  # Print the client version of odo
  odo version

Flags:
      --client   Client version only (no server required).
  -h, --help     Help for version

Additional Flags:
  -v, --v Level              Log level for V logs. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging

▶ ./odo service list --help        
List all services in the current application

Usage:
  odo service list [flags]

Examples:
  # List all services in the application
  odo service list

Flags:
      --app string       Application, defaults to active application
      --context string   Use given context directory as a source for component settings
  -h, --help             Help for list
      --project string   Project, defaults to active project

Additional Flags:
  -o, --o string             Specify output format, supported format: json (default "json")
  -v, --v Level              Log level for V logs. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
```

kadel commented:
One last think. It still shows up as an available global flag. Not a huge deal, but it would be great if we could hide it from there also.

cdrage commented:
@kadel 

Yes, that's the reason why I changed it from "Global Flags" to "Additional Flags" for each command as our libraries are dependent on having it under "Global". I also thought that the name "Additional Flags" would be better than Global for the sub-command / individual commands.

Plus I added it to there as it's nice to separate it from the "normal" flags. 

Oh! Or do you mean from `./odo --help`? I can remove it from that.

kadel commented:
> Oh! Or do you mean from `./odo --help`? I can remove it from that.

yep, that is what I meant ;-) Sorry, I should have been more specific 😇

cdrage commented:
@kadel Updated so it no longer appears in `./odo --help` !

cdrage commented:
/retest

cdrage commented on pkg/odo/genericclioptions/runnable.go:
```diff
@@ -30,15 +30,30 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) {
+	if hasFlagChanged && outputFlag.Value.String() == "json" && machineOutput == "" {
+
+		// By default we "disable" logging, so undisable it so that the below error can be shown.
+		flag.Set("o", "")
```
Which is fine :)

cdrage commented on pkg/odo/cli/cli.go:
```diff
@@ -134,12 +134,17 @@ func NewCmdOdo(name, fullName string) *cobra.Command {
 
+	// We will mark the command as hidden and then re-enable if the command
+	// supports json output
+	pflag.CommandLine.MarkHidden("o")
```
Which is fine :)

cdrage commented:
/retest

mik-dass commented:
```
'error: unable to connect to a server to handle {"image.openshift.io" "v1" "imagestreams"}: invalid configuration: Missing or incomplete configuration info.  Please login or point to an existing, complete config file:
[oc] 
[oc]   1. Via the command-line flag --config
[oc]   2. Via the KUBECONFIG environment variable
[oc]   3. In your home directory as ~/.kube/config
```
/retest

mik-dass commented on pkg/odo/cli/component/create.go:
```diff
@@ -593,7 +594,6 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
 	componentCreateCmd.Flags().StringSliceVarP(&co.componentPorts, "port", "p", []string{}, "Ports to be used when the component is created (ex. 8080,8100/tcp,9100/udp)")
 	componentCreateCmd.Flags().StringSliceVar(&co.componentEnvVars, "env", []string{}, "Environmental variables for the component. For example --env VariableName=Value")
 	// Add a defined annotation in order to appear in the help menu
```
please remove this comment

mik-dass commented on pkg/odo/cli/component/delete.go:
```diff
@@ -133,7 +134,6 @@ func NewCmdDelete(name, fullName string) *cobra.Command {
 	componentDeleteCmd.Flags().BoolVarP(&do.componentDeleteAllFlag, "all", "a", false, "Delete component and local config")
 
 	// Add a defined annotation in order to appear in the help menu
```
please remove this comment

mik-dass commented on pkg/odo/cli/component/describe.go:
```diff
@@ -88,18 +88,18 @@ func NewCmdDescribe(name, fullName string) *cobra.Command {
 		},
 	}
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/link.go:
```diff
@@ -134,7 +135,6 @@ func NewCmdLink(name, fullName string) *cobra.Command {
 	linkCmd.PersistentFlags().BoolVar(&o.waitForTarget, "wait-for-target", false, "If enabled, the link command will wait for the service to be provisioned (has no effect when linking to a component)")
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/list.go:
```diff
@@ -149,17 +149,17 @@ func NewCmdList(name, fullName string) *cobra.Command {
 			genericclioptions.GenericRun(o, cmd, args)
 		},
 	}
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/log.go:
```diff
@@ -74,7 +75,6 @@ func NewCmdLog(name, fullName string) *cobra.Command {
 	logCmd.Flags().BoolVarP(&o.logFollow, "follow", "f", false, "Follow logs")
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/push.go:
```diff
@@ -127,7 +128,6 @@ func NewCmdPush(name, fullName string) *cobra.Command {
 	pushCmd.Flags().BoolVarP(&po.forceBuild, "force-build", "f", false, "Use force-build flag to force building the component")
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/unlink.go:
```diff
@@ -87,7 +88,6 @@ func NewCmdUnlink(name, fullName string) *cobra.Command {
 	unlinkCmd.PersistentFlags().BoolVarP(&o.wait, "wait", "w", false, "If enabled the link will return only when the component is fully running after the link is deleted")
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/update.go:
```diff
@@ -167,7 +168,6 @@ func NewCmdUpdate(name, fullName string) *cobra.Command {
 	updateCmd.Flags().StringVarP(&uo.local, "local", "l", "", "Use local directory as a source for component.")
 	updateCmd.Flags().StringVarP(&uo.ref, "ref", "r", "", "Use a specific ref e.g. commit, branch or tag of the git repository")
 	// Add a defined annotation in order to appear in the help menu
```
same as above

mik-dass commented on pkg/odo/cli/component/watch.go:
```diff
@@ -162,7 +163,6 @@ func NewCmdWatch(name, fullName string) *cobra.Command {
 	watchCmd.Flags().IntVar(&wo.delay, "delay", 1, "Time in seconds between a detection of code change and push.delay=0 means changes will be pushed as soon as they are detected which can cause performance issues")
 
 	// Add a defined annotation in order to appear in the help menu
```
same as above

cdrage commented:
Thanks @mik-dass for the review! updated the PR again.

mik-dass commented:
```
agged from registry.centos.org/dotnet/dotnet-20-centos7
[oc] 
[oc]   ! error: Import failed (InternalError): Internal error occurred: Get https://registry.centos.org/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[oc]       Less than a second ago
[oc] 
```
/retest

girishramnani commented:
reviewed the PR just waiting to confirm if all the comments from @mik-dass were resolved

cdrage commented:
/kind bug
